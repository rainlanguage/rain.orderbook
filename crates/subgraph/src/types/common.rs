use crate::schema;
use serde::{Deserialize, Serialize};
use typeshare::typeshare;

#[derive(cynic::QueryVariables, Debug, Clone)]
#[typeshare]
pub struct IdQueryVariables<'a> {
    pub id: &'a cynic::Id,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrdersListFilterArgs {
    pub owners: Vec<Bytes>,
    pub active: Option<bool>,
    pub order_hash: Option<Bytes>,
}

#[derive(cynic::QueryVariables, Debug, Clone)]
#[typeshare]
pub struct PaginationQueryVariables {
    pub first: Option<i32>,
    pub skip: Option<i32>,
}

#[derive(cynic::InputObject, Debug, Clone)]
#[cynic(graphql_type = "Order_filter")]
#[typeshare]
pub struct OrdersListQueryFilters {
    #[cynic(rename = "owner_in", skip_serializing_if = "Vec::is_empty")]
    pub owner_in: Vec<Bytes>,
    #[cynic(rename = "active", skip_serializing_if = "Option::is_none")]
    pub active: Option<bool>,
    #[cynic(rename = "orderHash", skip_serializing_if = "Option::is_none")]
    pub order_hash: Option<Bytes>,
}

#[derive(cynic::QueryVariables, Debug, Clone)]
#[typeshare]
pub struct OrdersListQueryVariables {
    pub first: Option<i32>,
    pub skip: Option<i32>,
    #[cynic(rename = "filters")]
    pub filters: Option<OrdersListQueryFilters>,
}

#[derive(cynic::QueryVariables, Debug, Clone)]
#[typeshare]
pub struct PaginationWithIdQueryVariables {
    pub first: Option<i32>,
    pub id: Bytes,
    pub skip: Option<i32>,
}

#[derive(cynic::QueryVariables, Debug, Clone)]
#[typeshare]
pub struct PaginationWithTimestampQueryVariables {
    pub first: Option<i32>,
    pub id: Bytes,
    pub skip: Option<i32>,
    pub timestamp_gte: Option<BigInt>,
    pub timestamp_lte: Option<BigInt>,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[typeshare]
pub struct Orderbook {
    pub id: Bytes,
}

#[typeshare]
pub type RainMetaV1 = Bytes;

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct Order {
    pub id: Bytes,
    pub order_bytes: Bytes,
    pub order_hash: Bytes,
    pub owner: Bytes,
    pub outputs: Vec<Vault>,
    pub inputs: Vec<Vault>,
    pub orderbook: Orderbook,
    pub active: bool,
    pub timestamp_added: BigInt,
    pub meta: Option<RainMetaV1>,
    pub add_events: Vec<AddOrder>,
    pub trades: Vec<OrderStructPartialTrade>,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[cynic(graphql_type = "Order")]
#[serde(rename_all = "camelCase")]
#[typeshare]
pub struct TradeStructPartialOrder {
    pub id: Bytes,
    pub order_hash: Bytes,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[cynic(graphql_type = "Order")]
#[serde(rename_all = "camelCase")]
#[typeshare]
pub struct OrderAsIO {
    pub id: Bytes,
    pub order_hash: Bytes,
    pub active: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VaultsListFilterArgs {
    pub owners: Vec<Bytes>,
    pub hide_zero_balance: bool,
}

#[derive(cynic::InputObject, Debug, Clone)]
#[cynic(graphql_type = "Vault_filter")]
#[typeshare]
pub struct VaultsListQueryFilters {
    #[cynic(rename = "owner_in", skip_serializing_if = "Vec::is_empty")]
    pub owner_in: Vec<Bytes>,
    #[cynic(rename = "balance_gt", skip_serializing_if = "Option::is_none")]
    pub balance_gt: Option<BigInt>,
}

#[derive(cynic::QueryVariables, Debug, Clone)]
#[typeshare]
pub struct VaultsListQueryVariables {
    pub first: Option<i32>,
    pub skip: Option<i32>,
    #[cynic(rename = "filters")]
    pub filters: Option<VaultsListQueryFilters>,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct Vault {
    pub id: Bytes,
    pub owner: Bytes,
    pub vault_id: BigInt,
    pub balance: BigInt,
    pub token: Erc20,
    pub orderbook: Orderbook,
    // latest orders
    #[arguments(orderBy: timestampAdded, orderDirection: desc)]
    pub orders_as_output: Vec<OrderAsIO>,
    // latest orders
    #[arguments(orderBy: timestampAdded, orderDirection: desc)]
    pub orders_as_input: Vec<OrderAsIO>,
    pub balance_changes: Vec<VaultBalanceChange>,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "Vault")]
#[typeshare]
pub struct VaultBalanceChangeVault {
    pub id: Bytes,
    pub vault_id: BigInt,
    pub token: Erc20,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "VaultBalanceChange")]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct VaultBalanceChangeUnwrapped {
    pub __typename: String,
    pub amount: BigInt,
    pub new_vault_balance: BigInt,
    pub old_vault_balance: BigInt,
    pub vault: VaultBalanceChangeVault,
    pub timestamp: BigInt,
    pub transaction: Transaction,
    pub orderbook: Orderbook,
}

#[derive(cynic::InlineFragments, Debug, Clone, Serialize)]
#[serde(tag = "__typename", content = "data")]
#[serde(rename_all = "camelCase")]
#[typeshare]
pub enum VaultBalanceChange {
    Withdrawal(Withdrawal),
    TradeVaultBalanceChange(TradeVaultBalanceChange),
    Deposit(Deposit),
    ClearBounty(ClearBounty),
    #[cynic(fallback)]
    Unknown,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct Deposit {
    pub id: Bytes,
    pub __typename: String,
    pub amount: BigInt,
    pub new_vault_balance: BigInt,
    pub old_vault_balance: BigInt,
    pub vault: VaultBalanceChangeVault,
    pub timestamp: BigInt,
    pub transaction: Transaction,
    pub orderbook: Orderbook,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct Withdrawal {
    pub id: Bytes,
    pub __typename: String,
    pub amount: BigInt,
    pub new_vault_balance: BigInt,
    pub old_vault_balance: BigInt,
    pub vault: VaultBalanceChangeVault,
    pub timestamp: BigInt,
    pub transaction: Transaction,
    pub orderbook: Orderbook,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct TradeVaultBalanceChange {
    pub id: Bytes,
    pub __typename: String,
    pub amount: BigInt,
    pub new_vault_balance: BigInt,
    pub old_vault_balance: BigInt,
    pub vault: VaultBalanceChangeVault,
    pub timestamp: BigInt,
    pub transaction: Transaction,
    pub orderbook: Orderbook,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct ClearBounty {
    pub id: Bytes,
    pub __typename: String,
    pub amount: BigInt,
    pub new_vault_balance: BigInt,
    pub old_vault_balance: BigInt,
    pub vault: VaultBalanceChangeVault,
    pub timestamp: BigInt,
    pub transaction: Transaction,
    pub orderbook: Orderbook,
    pub sender: Bytes,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[typeshare]
pub struct TradeEvent {
    pub transaction: Transaction,
    pub sender: Bytes,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct Trade {
    pub id: Bytes,
    pub trade_event: TradeEvent,
    pub output_vault_balance_change: TradeVaultBalanceChange,
    pub order: TradeStructPartialOrder,
    pub input_vault_balance_change: TradeVaultBalanceChange,
    pub timestamp: BigInt,
    pub orderbook: Orderbook,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "Trade")]
#[typeshare]
pub struct OrderStructPartialTrade {
    pub id: Bytes,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone, PartialEq)]
#[cynic(graphql_type = "ERC20")]
#[typeshare]
pub struct Erc20 {
    pub id: Bytes,
    pub address: Bytes,
    pub name: Option<String>,
    pub symbol: Option<String>,
    pub decimals: Option<BigInt>,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[typeshare]
#[serde(rename_all = "camelCase")]
pub struct Transaction {
    pub id: Bytes,
    pub from: Bytes,
    pub block_number: BigInt,
    pub timestamp: BigInt,
}

#[derive(cynic::QueryFragment, Debug, Serialize, Clone)]
#[typeshare]
pub struct AddOrder {
    pub transaction: Transaction,
}

#[derive(cynic::Scalar, Debug, Clone, PartialEq)]
#[typeshare]
pub struct BigInt(pub String);

#[derive(cynic::Scalar, Debug, Clone, PartialEq)]
#[typeshare]
pub struct Bytes(pub String);

#[derive(cynic::Enum, Clone, Copy, Debug)]
#[typeshare]
pub enum OrderDirection {
    #[cynic(rename = "asc")]
    Asc,
    #[cynic(rename = "desc")]
    Desc,
}

#[derive(cynic::Enum, Clone, Copy, Debug)]
#[cynic(graphql_type = "Order_orderBy")]
#[typeshare]
pub enum OrderOrderBy {
    #[cynic(rename = "id")]
    Id,
    #[cynic(rename = "orderbook")]
    Orderbook,
    #[cynic(rename = "orderbook__id")]
    OrderbookId,
    #[cynic(rename = "active")]
    Active,
    #[cynic(rename = "orderHash")]
    OrderHash,
    #[cynic(rename = "owner")]
    Owner,
    #[cynic(rename = "inputs")]
    Inputs,
    #[cynic(rename = "outputs")]
    Outputs,
    #[cynic(rename = "nonce")]
    Nonce,
    #[cynic(rename = "orderBytes")]
    OrderBytes,
    #[cynic(rename = "addEvents")]
    AddEvents,
    #[cynic(rename = "removeEvents")]
    RemoveEvents,
    #[cynic(rename = "trades")]
    Trades,
    #[cynic(rename = "meta")]
    Meta,
    #[cynic(rename = "timestampAdded")]
    TimestampAdded,
}

#[derive(cynic::Enum, Clone, Copy, Debug)]
#[cynic(graphql_type = "AddOrder_orderBy")]
#[typeshare]
pub enum AddOrderOrderBy {
    #[cynic(rename = "id")]
    Id,
    #[cynic(rename = "order")]
    Order,
    #[cynic(rename = "order__id")]
    OrderId,
    #[cynic(rename = "order__active")]
    OrderActive,
    #[cynic(rename = "order__orderHash")]
    OrderOrderHash,
    #[cynic(rename = "order__owner")]
    OrderOwner,
    #[cynic(rename = "order__nonce")]
    OrderNonce,
    #[cynic(rename = "order__orderBytes")]
    OrderOrderBytes,
    #[cynic(rename = "order__meta")]
    OrderMeta,
    #[cynic(rename = "order__timestampAdded")]
    OrderTimestampAdded,
    #[cynic(rename = "orderbook")]
    Orderbook,
    #[cynic(rename = "orderbook__id")]
    OrderbookId,
    #[cynic(rename = "transaction")]
    Transaction,
    #[cynic(rename = "transaction__id")]
    TransactionId,
    #[cynic(rename = "transaction__timestamp")]
    TransactionTimestamp,
    #[cynic(rename = "transaction__blockNumber")]
    TransactionBlockNumber,
    #[cynic(rename = "transaction__from")]
    TransactionFrom,
    #[cynic(rename = "sender")]
    Sender,
}

#[derive(cynic::Enum, Clone, Copy, Debug)]
#[cynic(graphql_type = "Trade_orderBy")]
#[typeshare]
pub enum TradeOrderBy {
    #[cynic(rename = "id")]
    Id,
    #[cynic(rename = "orderbook")]
    Orderbook,
    #[cynic(rename = "orderbook__id")]
    OrderbookId,
    #[cynic(rename = "order")]
    Order,
    #[cynic(rename = "order__id")]
    OrderId,
    #[cynic(rename = "order__active")]
    OrderActive,
    #[cynic(rename = "order__orderHash")]
    OrderOrderHash,
    #[cynic(rename = "order__owner")]
    OrderOwner,
    #[cynic(rename = "order__nonce")]
    OrderNonce,
    #[cynic(rename = "order__orderBytes")]
    OrderOrderBytes,
    #[cynic(rename = "order__meta")]
    OrderMeta,
    #[cynic(rename = "order__timestampAdded")]
    OrderTimestampAdded,
    #[cynic(rename = "inputVaultBalanceChange")]
    InputVaultBalanceChange,
    #[cynic(rename = "inputVaultBalanceChange__id")]
    InputVaultBalanceChangeId,
    #[cynic(rename = "inputVaultBalanceChange__amount")]
    InputVaultBalanceChangeAmount,
    #[cynic(rename = "inputVaultBalanceChange__oldVaultBalance")]
    InputVaultBalanceChangeOldVaultBalance,
    #[cynic(rename = "inputVaultBalanceChange__newVaultBalance")]
    InputVaultBalanceChangeNewVaultBalance,
    #[cynic(rename = "inputVaultBalanceChange__timestamp")]
    InputVaultBalanceChangeTimestamp,
    #[cynic(rename = "outputVaultBalanceChange")]
    OutputVaultBalanceChange,
    #[cynic(rename = "outputVaultBalanceChange__id")]
    OutputVaultBalanceChangeId,
    #[cynic(rename = "outputVaultBalanceChange__amount")]
    OutputVaultBalanceChangeAmount,
    #[cynic(rename = "outputVaultBalanceChange__oldVaultBalance")]
    OutputVaultBalanceChangeOldVaultBalance,
    #[cynic(rename = "outputVaultBalanceChange__newVaultBalance")]
    OutputVaultBalanceChangeNewVaultBalance,
    #[cynic(rename = "outputVaultBalanceChange__timestamp")]
    OutputVaultBalanceChangeTimestamp,
    #[cynic(rename = "tradeEvent")]
    TradeEvent,
    #[cynic(rename = "tradeEvent__id")]
    TradeEventId,
    #[cynic(rename = "tradeEvent__sender")]
    TradeEventSender,
    #[cynic(rename = "timestamp")]
    Timestamp,
}

#[derive(cynic::Enum, Clone, Copy, Debug)]
#[cynic(graphql_type = "Vault_orderBy")]
#[typeshare]
pub enum VaultOrderBy {
    #[cynic(rename = "id")]
    Id,
    #[cynic(rename = "orderbook")]
    Orderbook,
    #[cynic(rename = "orderbook__id")]
    OrderbookId,
    #[cynic(rename = "token")]
    Token,
    #[cynic(rename = "token__id")]
    TokenId,
    #[cynic(rename = "token__address")]
    TokenAddress,
    #[cynic(rename = "token__name")]
    TokenName,
    #[cynic(rename = "token__symbol")]
    TokenSymbol,
    #[cynic(rename = "token__decimals")]
    TokenDecimals,
    #[cynic(rename = "owner")]
    Owner,
    #[cynic(rename = "vaultId")]
    VaultId,
    #[cynic(rename = "ordersAsInput")]
    OrdersAsInput,
    #[cynic(rename = "ordersAsOutput")]
    OrdersAsOutput,
    #[cynic(rename = "balance")]
    Balance,
    #[cynic(rename = "balanceChanges")]
    BalanceChanges,
}
