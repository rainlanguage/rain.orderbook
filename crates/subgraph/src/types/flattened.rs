use std::num::ParseIntError;

use super::{
    orders_list, vault_balance_change::VaultBalanceChange, vault_list_balance_changes, vaults_list,
};
use crate::utils::format_bigint_timestamp_display;
use crate::{csv::WriteCsv, utils::FormatTimestampDisplayError};
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum TryIntoFlattenedError {
    #[error(transparent)]
    FormatTimestampDisplayError(#[from] FormatTimestampDisplayError),
    #[error(transparent)]
    ParseIntError(#[from] ParseIntError),
}

#[derive(Serialize, Deserialize, Clone)]
pub struct TokenVaultFlattened {
    pub id: String,
    pub owner: vaults_list::Bytes,
    pub vault_id: vaults_list::BigInt,
    pub token_name: String,
    pub token_symbol: String,
    pub token_decimals: i32,
    pub token_address: String,
    pub balance_display: vaults_list::BigDecimal,
    pub balance: vaults_list::BigInt,
}

impl From<vaults_list::TokenVault> for TokenVaultFlattened {
    fn from(val: vaults_list::TokenVault) -> Self {
        Self {
            id: val.id.into_inner(),
            owner: val.owner.id,
            vault_id: val.vault_id,
            token_name: val.token.name,
            token_symbol: val.token.symbol,
            token_decimals: val.token.decimals,
            token_address: val.token.id.into_inner(),
            balance_display: val.balance_display,
            balance: val.balance,
        }
    }
}

impl WriteCsv<TokenVaultFlattened> for Vec<TokenVaultFlattened> {}

#[derive(Serialize, Deserialize, Clone)]
pub struct OrderFlattened {
    pub id: String,
    pub timestamp: orders_list::BigInt,
    pub timestamp_display: String,
    pub handle_io: bool,
    pub owner: orders_list::Bytes,
    pub order_active: bool,
    pub interpreter: orders_list::Bytes,
    pub interpreter_store: orders_list::Bytes,
    pub transaction: String,
    pub valid_inputs_vaults: String,
    pub valid_outputs_vaults: String,
    pub valid_inputs_token_symbols_display: String,
    pub valid_outputs_token_symbols_display: String,
}

impl TryFrom<orders_list::Order> for OrderFlattened {
    type Error = TryIntoFlattenedError;

    fn try_from(val: orders_list::Order) -> Result<Self, Self::Error> {
        Ok(Self {
            id: val.id.into_inner(),
            timestamp: val.timestamp.clone(),
            timestamp_display: format_bigint_timestamp_display(val.timestamp.0)?,
            handle_io: val.handle_io,
            owner: val.owner.id,
            order_active: val.order_active,
            interpreter: val.interpreter,
            interpreter_store: val.interpreter_store,
            transaction: val.transaction.id.into_inner(),
            valid_inputs_vaults: val.valid_inputs.clone().map_or(
                "".into(),
                |v: Vec<orders_list::Io>| {
                    v.into_iter()
                        .map(|io| io.token_vault.id.into_inner())
                        .collect::<Vec<String>>()
                        .join(", ")
                },
            ),
            valid_outputs_vaults: val.valid_outputs.clone().map_or("".into(), |v| {
                v.into_iter()
                    .map(|io| io.token_vault.id.into_inner())
                    .collect::<Vec<String>>()
                    .join(", ")
            }),
            valid_inputs_token_symbols_display: val.valid_inputs.map_or(
                "".into(),
                |v: Vec<orders_list::Io>| {
                    v.into_iter()
                        .map(|io| io.token.symbol)
                        .collect::<Vec<String>>()
                        .join(", ")
                },
            ),
            valid_outputs_token_symbols_display: val.valid_outputs.map_or("".into(), |v| {
                v.into_iter()
                    .map(|io| io.token.symbol)
                    .collect::<Vec<String>>()
                    .join(", ")
            }),
        })
    }
}

impl WriteCsv<OrderFlattened> for Vec<OrderFlattened> {}

#[derive(Serialize, Deserialize, Clone)]
pub struct VaultBalanceChangeFlattened {
    pub id: String,
    pub timestamp: vault_list_balance_changes::BigInt,
    pub timestamp_display: String,
    pub sender: vault_list_balance_changes::Bytes,
    pub amount: vault_list_balance_changes::BigDecimal,
    pub amount_display_signed: String,
    pub change_type_display: String,
    pub balance: vault_list_balance_changes::BigDecimal,
}

impl TryFrom<VaultBalanceChange> for VaultBalanceChangeFlattened {
    type Error = TryIntoFlattenedError;

    fn try_from(val: VaultBalanceChange) -> Result<Self, Self::Error> {
        match val {
            VaultBalanceChange::Deposit(v) => Ok(Self {
                id: v.id.into_inner(),
                timestamp: v.timestamp.clone(),
                timestamp_display: format_bigint_timestamp_display(v.timestamp.0)?,
                sender: v.sender.id,
                amount: v.amount_display.clone(),
                amount_display_signed: v.amount_display.0,
                change_type_display: String::from("Deposit"),
                balance: v.token_vault.balance_display,
            }),
            VaultBalanceChange::Withdraw(v) => Ok(Self {
                id: v.id.into_inner(),
                timestamp: v.timestamp.clone(),
                timestamp_display: format_bigint_timestamp_display(v.timestamp.0)?,
                sender: v.sender.id,
                amount: v.amount_display.clone(),
                amount_display_signed: format!("-{}", v.amount_display.0),
                change_type_display: String::from("Withdraw"),
                balance: v.token_vault.balance_display,
            }),
        }
    }
}

impl WriteCsv<VaultBalanceChangeFlattened> for Vec<VaultBalanceChangeFlattened> {}
