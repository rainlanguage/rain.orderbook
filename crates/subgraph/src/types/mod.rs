pub mod common;
pub mod order;
pub mod order_detail_traits;
pub mod order_trade;
pub mod vault;

pub use cynic::Id;
