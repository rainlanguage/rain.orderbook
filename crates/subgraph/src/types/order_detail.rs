use crate::schema;
use serde::Serialize;
use typeshare::typeshare;

#[typeshare]
#[derive(cynic::QueryVariables, Debug)]
pub struct OrderDetailQueryVariables<'a> {
    pub id: &'a cynic::Id,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "Query", variables = "OrderDetailQueryVariables")]
pub struct OrderDetailQuery {
    #[arguments(id: $id)]
    pub order: Option<Order>,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
pub struct Order {
    pub id: cynic::Id,
    pub owner: Account,
    pub order_active: bool,
    pub interpreter: Bytes,
    pub interpreter_store: Bytes,
    pub expression_deployer: Bytes,
    pub expression: Bytes,
    pub timestamp: BigInt,
    #[cynic(rename = "handleIO")]
    pub handle_io: bool,
    pub valid_inputs: Option<Vec<Io>>,
    pub valid_outputs: Option<Vec<Io>>,
    pub meta: Option<RainMetaV1>,
    #[cynic(rename = "orderJSONString")]
    pub order_json: String,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
pub struct RainMetaV1 {
    pub meta_bytes: Bytes,
    pub content: Vec<ContentMetaV1>,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "IO")]
pub struct Io {
    pub token_vault: TokenVault,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
pub struct TokenVault {
    pub id: cynic::Id,
    pub vault_id: BigInt,
    pub vault: Vault,
    pub token: Erc20,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
pub struct Vault {
    pub owner: Account,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "ERC20")]
pub struct Erc20 {
    pub id: cynic::Id,
    pub name: String,
    pub symbol: String,
    pub decimals: i32,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
pub struct ContentMetaV1 {
    pub id: Bytes,
    pub payload: Bytes,
    pub magic_number: BigInt,
    pub content_type: Option<String>,
    pub content_encoding: Option<String>,
    pub content_language: Option<String>,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
pub struct Account {
    pub id: Bytes,
}

#[derive(cynic::Scalar, Debug, Clone)]
pub struct BigInt(pub String);

#[derive(cynic::Scalar, Debug, Clone)]
pub struct Bytes(pub String);
