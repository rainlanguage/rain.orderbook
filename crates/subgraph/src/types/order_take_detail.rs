use crate::schema;
use serde::Serialize;
use typeshare::typeshare;

#[derive(cynic::QueryVariables, Debug)]
pub struct OrderTakeDetailQueryVariables<'a> {
    pub id: &'a cynic::Id,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "OrderTakeDetailQueryVariables")]
pub struct OrderTakeDetailQuery {
    #[arguments(id: $id)]
    pub take_order_entity: Option<TakeOrderEntity>,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct TakeOrderEntity {
    pub id: cynic::Id,
    pub transaction: Transaction,
    pub sender: Account,
    pub timestamp: BigInt,
    pub order: Order,
    #[cynic(rename = "IORatio")]
    pub ioratio: BigDecimal,
    pub input: BigInt,
    pub input_display: BigDecimal,
    pub input_token: Erc20,
    #[cynic(rename = "inputIOIndex")]
    pub input_ioindex: BigInt,
    pub output: BigInt,
    pub output_display: BigDecimal,
    pub output_token: Erc20,
    #[cynic(rename = "outputIOIndex")]
    pub output_ioindex: BigInt,
    pub context: Option<ContextEntity>,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct Transaction {
    pub id: cynic::Id,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct Order {
    pub id: cynic::Id,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Serialize)]
#[cynic(graphql_type = "ERC20")]
pub struct Erc20 {
    pub id: cynic::Id,
    pub name: String,
    pub symbol: String,
    pub decimals: i32,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct ContextEntity {
    pub calling_context: Option<Vec<BigInt>>,
    pub calculations_context: Option<Vec<BigInt>>,
    pub vault_inputs_context: Option<Vec<BigInt>>,
    pub vault_outputs_context: Option<Vec<BigInt>>,
}

#[typeshare]
#[derive(cynic::QueryFragment, Debug, Serialize)]
pub struct Account {
    pub id: Bytes,
}

#[derive(cynic::Scalar, Debug, Clone)]
pub struct BigDecimal(pub String);

#[derive(cynic::Scalar, Debug, Clone)]
pub struct BigInt(pub String);

#[derive(cynic::Scalar, Debug, Clone)]
pub struct Bytes(pub String);
