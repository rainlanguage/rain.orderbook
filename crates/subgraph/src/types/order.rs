use super::common::*;
use crate::schema;
use serde::Serialize;
use typeshare::typeshare;

#[derive(cynic::QueryVariables, Debug)]
#[typeshare]
pub struct BatchOrderDetailQueryVariables {
    #[cynic(rename = "id_list")]
    pub id_list: OrderIdList,
}

#[derive(cynic::InputObject, Debug, Clone)]
#[cynic(graphql_type = "Order_filter")]
#[typeshare]
pub struct OrderIdList {
    #[cynic(rename = "id_in")]
    pub id_in: Vec<Bytes>,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "Query", variables = "BatchOrderDetailQueryVariables")]
#[typeshare]
pub struct BatchOrderDetailQuery {
    #[typeshare(typescript(type = "OrderSubgraph[]"))]
    #[arguments(where: $id_list)]
    pub orders: Vec<Order>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "OrdersListQueryVariables")]
#[typeshare]
pub struct OrdersListQuery {
    #[typeshare(typescript(type = "OrderSubgraph[]"))]
    #[arguments(orderBy: "timestampAdded", orderDirection: "desc", skip: $skip, first: $first, where: $filters)]
    pub orders: Vec<Order>,
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Query", variables = "IdQueryVariables")]
#[typeshare]
pub struct OrderDetailQuery {
    #[arguments(id: $id)]
    #[typeshare(typescript(type = "OrderSubgraph"))]
    pub order: Option<Order>,
}
