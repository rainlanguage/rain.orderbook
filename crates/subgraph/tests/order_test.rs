use cynic::Id;
use insta::assert_snapshot;
use rain_orderbook_subgraph_client::types::common::*;
use rain_orderbook_subgraph_client::types::order::SgOrderDetailByIdQuery;

#[test]
fn orders_query_gql_output() {
    use cynic::QueryBuilder;

    let id = Id::new("1234");
    let request_body = SgOrderDetailByIdQuery::build(SgIdQueryVariables { id: &id });

    assert_snapshot!(request_body.query);
}
