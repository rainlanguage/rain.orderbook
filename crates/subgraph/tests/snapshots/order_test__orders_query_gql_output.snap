---
source: crates/subgraph/tests/order_test.rs
expression: request_body.query
---
query OrderDetailQuery($id: ID!) {
  order(id: $id) {
    id
    orderBytes
    orderHash
    owner
    outputs {
      id
      owner
      vaultId
      balance
      token {
        id
        address
        name
        symbol
        decimals
      }
      orderbook {
        id
      }
      ordersAsOutput(orderBy: timestampAdded, orderDirection: desc) {
        id
        orderHash
        active
      }
      ordersAsInput(orderBy: timestampAdded, orderDirection: desc) {
        id
        orderHash
        active
      }
      balanceChanges {
        __typename
        ... on Withdrawal {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
        }
        ... on TradeVaultBalanceChange {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
        }
        ... on Deposit {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
        }
        ... on ClearBounty {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
          sender
        }
      }
    }
    inputs {
      id
      owner
      vaultId
      balance
      token {
        id
        address
        name
        symbol
        decimals
      }
      orderbook {
        id
      }
      ordersAsOutput(orderBy: timestampAdded, orderDirection: desc) {
        id
        orderHash
        active
      }
      ordersAsInput(orderBy: timestampAdded, orderDirection: desc) {
        id
        orderHash
        active
      }
      balanceChanges {
        __typename
        ... on Withdrawal {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
        }
        ... on TradeVaultBalanceChange {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
        }
        ... on Deposit {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
        }
        ... on ClearBounty {
          id
          __typename
          amount
          newVaultBalance
          oldVaultBalance
          vault {
            id
            vaultId
            token {
              id
              address
              name
              symbol
              decimals
            }
          }
          timestamp
          transaction {
            id
            from
            blockNumber
            timestamp
          }
          orderbook {
            id
          }
          sender
        }
      }
    }
    orderbook {
      id
    }
    active
    timestampAdded
    meta
    addEvents {
      transaction {
        id
        from
        blockNumber
        timestamp
      }
    }
    trades {
      id
    }
  }
}
